examples/scale/scenarios/backfill/backfill.go

```diff
@@ -0,0 +1,254 @@
+				},
+			},
+		},
+	}
+}
+
+func (s *BackfillScenario) Ticket() *pb.Ticket {
+	return &pb.Ticket{}
+}
+
+func (s *BackfillScenario) Backfill() *pb.Backfill {
+	return &pb.Backfill{}
+}
+
+func (s *BackfillScenario) MatchFunction(p *pb.MatchProfile, poolBackfills map[string][]*pb.Backfill, poolTickets map[string][]*pb.Ticket) ([]*pb.Match, error) {
+	return statefullMMF(p, poolBackfills, poolTickets, s.TicketsPerMatch, s.MaxTicketsPerNotFullMatch)
+}
+
+func statefullMMF(p *pb.MatchProfile, poolBackfills map[string][]*pb.Backfill, poolTickets map[string][]*pb.Ticket, ticketsPerMatch int, maxTicketsPerNotFullMatch int) ([]*pb.Match, error) {
+	var matches []*pb.Match
+
+	for pool, backfills := range poolBackfills {
+		tickets, ok := poolTickets[pool]
+
+		if !ok || len(tickets) == 0 {
+			// no tickets in pool
+			continue
+		}
+
+		// process backfills first
+		for _, b := range backfills {
+			l := len(tickets)
+			if l == 0 {
+				// no tickets left
+				break
+			}
+
+			openSlots := getOpenSlots(b, ticketsPerMatch)
+			if openSlots <= 0 {
+				// no free open slots
+				continue
+			}
+
+			if l > openSlots {
+				l = openSlots
+			}
+
+			setOpenSlots(b, openSlots-l)
+			matches = append(matches, &pb.Match{
+				MatchId:       fmt.Sprintf("profile-%v-time-%v-%v", p.GetName(), time.Now().Format("2006-01-02T15:04:05.00"), len(matches)),
+				Tickets:       tickets[0:l],
+				MatchProfile:  p.GetName(),
+				MatchFunction: "backfill",
+				Backfill:      b,
+			})
+			tickets = tickets[l:]
+		}
+
+		// create not full matches with backfill
+		for {
+			l := len(tickets)
+			if l == 0 {
+				// no tickets left
+				break
+			}
+
+			if l > maxTicketsPerNotFullMatch {
+				l = maxTicketsPerNotFullMatch
+			}
+			b := pb.Backfill{}
+			setOpenSlots(&b, ticketsPerMatch-l)
+			matches = append(matches, &pb.Match{
+				MatchId:            fmt.Sprintf("profile-%v-time-%v-%v", p.GetName(), time.Now().Format("2006-01-02T15:04:05.00"), len(matches)),
+				Tickets:            tickets[0:l],
+				MatchProfile:       p.GetName(),
+				MatchFunction:      "backfill",
+				Backfill:           &b,
+				AllocateGameserver: true,
+			})
+			tickets = tickets[l:]
+		}
+	}
+
+	return matches, nil
+}
+
+func getOpenSlots(b *pb.Backfill, defaultVal int) int {
+	if b.Extensions == nil {
+		return defaultVal
+	}
+
+	any, ok := b.Extensions[openSlotsKey]
+	if !ok {
+		return defaultVal
+	}
+
+	var val wrappers.Int32Value
+	err := ptypes.UnmarshalAny(any, &val)
+	if err != nil {
+		panic(err)
+	}
+
+	return int(val.Value)
+}
+
+func setOpenSlots(b *pb.Backfill, val int) {
+	if b.Extensions == nil {
+		b.Extensions = make(map[string]*any.Any)
+	}
+
+	any, err := ptypes.MarshalAny(&wrappers.Int32Value{Value: int32(val)})
+	if err != nil {
+		panic(err)
+	}
+
+	b.Extensions[openSlotsKey] = any
+}
+
+func statelessMMF(p *pb.MatchProfile, poolBackfills map[string][]*pb.Backfill, poolTickets map[string][]*pb.Ticket, ticketsPerMatch int) ([]*pb.Match, error) {
```

Comment:
What's this method for?  Not used anywhere?

Reply:
I tried to implement 2 different backfill scenarios:
1. N backfills created initially by frontend and MMF takes that backfills and fills them. After that backfills are acknowledged by backend. (statelessMMF)
2. MMF creates backfill and fill it with N tickets on the first FetchMatches. Updates backfill's extensions with remaining open slots (= K - N where K - number of tickets for full match). After that backfill is acknowledged by backend. On the second FetchMatches MMF takes previously created backfill and tries to fill it with remaining tickets. After that backfill is acknowledged and deleted by backend. (statefullMMF)

In order to use 1 or 2 scenario I should configure Scenario struct in scenarios.go and call proper MMF implementation in MatchFunction function.

Maybe namings statefullMMF and statelessMMF are not good here.